src/Tools/dotnet-monitor/Auth/ApiKeyAuthenticationHostedService.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor
+{
+    /// <summary>
+    /// Service that monitors API Key authentication options changes and logs issues with the specified options.
+    /// </summary>
+    internal class ApiKeyAuthenticationHostedService :
+        IHostedService,
+        IDisposable
+    {
+        private readonly ILogger<ApiKeyAuthenticationHostedService> _logger;
+        private readonly IOptionsMonitor<ApiKeyAuthenticationOptions> _options;
+
+        private IDisposable _changeRegistration;
+
+        public ApiKeyAuthenticationHostedService(
+            ILogger<ApiKeyAuthenticationHostedService> logger,
+            IOptionsMonitor<ApiKeyAuthenticationOptions> options
+            )
+        {
+            _logger = logger;
+            _options = options;
+        }
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            _changeRegistration = _options.OnChange(OnApiKeyAuthenticationOptionsChanged);
```

Comment:
I don't quite understand why we need a hosted service for this. Couldn't you
1) Check configuration after setting up the app to write out errors
2) At the same time register for change registration and then produce the same warnings?

Reply:
Exactly where else would I do that though? I could do that in the command handler, but that feels like throwing this there because it's convenient. The Startup class doesn't have a lifetime notion so I could unregistered the change token. 

Would creating it as a normal singleton that the Startup class initialized be acceptable? I'll have to see if the change token registration can be disposed on service disposal.

Reply:
Assuming a normal singleton's IDispose works, then I think that would be a better way to go.